traversal/Traversal.java

```diff
@@ -88,30 +95,54 @@ public int hashCode() {
     public static class Parameters {
 
         private final Map<Identifier.Variable, VertexIID.Thing> iids;
-        private final Map<Pair<Identifier.Variable, Predicate.Value<?, ?>>, Set<Value>> values;
+        private final Map<Pair<Identifier.Variable, Predicate.Value<?, ?>>, Set<Value<?>>> values;
+        private final Map<Identifier.Variable, Pair<Predicate.Value<?, ?>, Value<?>>> largestGTPredicates;
+        private final Map<Identifier.Variable, Pair<Predicate.Value<?, ?>, Value<?>>> smallestLTPredicates;
```

Comment:
we pre-compute the largest and smallest predicates for comparison, which we used for optimising predicates